File: backend/src/main/java/com/backend/ord/config/security/JwtFactory.java
package com.backend.ord.config.security;

import com.backend.ord.config.properties.JwtProperties;
import com.backend.ord.domain.entities.User;
import com.backend.ord.exceptions.UserNotFoundException;
import com.backend.ord.services.UserSessionService;
import com.backend.ord.utils.CookieUtils;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.stereotype.Component;

import java.util.Map;

@Component
public class JwtFactory {
    private final JwtService jwtService;
    private final JwtProperties jwtProperties;

    private final UserSessionService userSessionService;

    public JwtFactory(JwtService jwtService, JwtProperties jwtProperties, UserSessionService userSessionService) {
        this.jwtService = jwtService;
        this.jwtProperties = jwtProperties;
        this.userSessionService = userSessionService;
    }


    public String createTokenForUser(
            User user,
            HttpServletResponse response
    ) throws UserNotFoundException {
        // Generate token
        String token = jwtService.generateToken(
                getExtraClaims(user),
                user
        );

        // Create cookie
        createCookie(token, response);

        // Create new user session
        createNewUserSession(token);

        // Return generated JWT token
        return token;
    }

    // ### Helpers

    private Map<String, Object> getExtraClaims(User user) {
        String claimName = jwtProperties.getUserIdClaimName();

        return Map.of(
                claimName, user.getId()
        );
    }

    private void createCookie(String token, HttpServletResponse response) {
        String cookieName = jwtProperties.getAuthCookieName();
        CookieUtils.createCookie(cookieName, token, response);
    }

    private void createNewUserSession(String token) throws UserNotFoundException {
        userSessionService.openSessionFromJWT(token);
    }
}

File: backend/src/main/kotlin/com/backend/ord/config/security/JwtFactory.kt
package com.backend.ord.config.security

import com.backend.ord.config.properties.JwtProperties
import com.backend.ord.core.user.model.UserEntity
import com.backend.ord.exceptions.UserNotFoundException
import com.backend.ord.services.UserSessionService
import com.backend.ord.utils.CookieUtils.createCookie
import jakarta.servlet.http.HttpServletResponse
import org.springframework.stereotype.Component

@Component
class JwtFactory(
    private val jwtService: JwtService,
    private val jwtProperties: JwtProperties,
    private val userSessionService: UserSessionService
) {
    @Throws(UserNotFoundException::class)
    fun createTokenForUser(
        user: UserEntity,
        response: HttpServletResponse
    ): String? {
        // Generate token
        val token = jwtService.generateToken(
            extraClaims = getExtraClaims(user),
            userDetails = user
        ).apply {
            createCookie(
                token = this,
                response = response
            )

            createNewUserSession(token = this)
        }

        return token
    }

    // ### Helpers
    private fun getExtraClaims(user: UserEntity): Map<String, Any> {
        val claimName = jwtProperties.userIdClaimName
        val claimValue = user.id.toString()

        return mapOf(claimName to claimValue)
    }

    private fun createCookie(token: String, response: HttpServletResponse) {
        createCookie(
            name = jwtProperties.authCookieName,
            value = token,
            response = response
        )
    }

    @Throws(UserNotFoundException::class)
    private fun createNewUserSession(token: String) {
        userSessionService.openSessionFromJWT(token)
    }
}
